deps/patches/gmp_alloc_overflow_func.patch

```diff
@@ -0,0 +1,230 @@
++  do {                                                           \
++    (*__gmp_alloc_overflow_func) ();                             \
++    fprintf (stderr, "unexpected return from alloc_overflow\n"); \
++    abort ();                                                    \
++  } while (0)
+ 
+ /* Dummy for non-gcc, code involving it will go dead. */
+ #if ! defined (__GNUC__) || __GNUC__ < 2
+ #define __builtin_constant_p(x)   0
+ #endif
+diff --git a/memory.c b/memory.c
+--- a/memory.c
++++ b/memory.c
+@@ -36,10 +36,11 @@ see https://www.gnu.org/licenses/.  */
+ 
+ 
+ void * (*__gmp_allocate_func) (size_t) = __gmp_default_allocate;
+ void * (*__gmp_reallocate_func) (void *, size_t, size_t) = __gmp_default_reallocate;
+ void   (*__gmp_free_func) (void *, size_t) = __gmp_default_free;
++void   (*__gmp_alloc_overflow_func) () = __gmp_default_alloc_overflow;
+ 
+ 
+ /* Default allocation functions.  In case of failure to allocate/reallocate
+    an error message is written to stderr and the program aborts.  */
+ 
+@@ -142,5 +143,12 @@ void
+     blk_ptr = p - 1;
+   }
+ #endif
+   free (blk_ptr);
+ }
++
++void
++__gmp_default_alloc_overflow(void)
++{
++    fprintf(stderr, "__gmp_default_alloc_overflow called\n");
++    abort();
++}
+diff --git a/mp_get_fns.c b/mp_get_fns.c
+--- a/mp_get_fns.c
++++ b/mp_get_fns.c
+@@ -44,5 +44,13 @@ mp_get_memory_functions (void *(**alloc_
+     *realloc_func = __gmp_reallocate_func;
+ 
+   if (free_func != NULL)
+     *free_func = __gmp_free_func;
+ }
++
++void
++mp_get_alloc_overflow_function(
++        void (**alloc_overflow_func) (void)) __GMP_NOTHROW
++{
++  if (alloc_overflow_func != NULL)
++    *alloc_overflow_func = __gmp_alloc_overflow_func;
++}
+diff --git a/mp_set_fns.c b/mp_set_fns.c
+--- a/mp_set_fns.c
++++ b/mp_set_fns.c
+@@ -46,5 +46,13 @@ mp_set_memory_functions (void *(*alloc_f
+ 
+   __gmp_allocate_func = alloc_func;
+   __gmp_reallocate_func = realloc_func;
+   __gmp_free_func = free_func;
+ }
++
++mp_set_alloc_overflow_function(
++             void (*alloc_overflow_func) (void)) __GMP_NOTHROW
++{
++  if (alloc_overflow_func == 0)
++    alloc_overflow_func = __gmp_default_alloc_overflow;
++  __gmp_alloc_overflow_func = alloc_overflow_func;
++}
+diff --git a/mpz/init2.c b/mpz/init2.c
+--- a/mpz/init2.c
++++ b/mpz/init2.c
+@@ -44,11 +44,11 @@ mpz_init2 (mpz_ptr x, mp_bitcnt_t bits)
+   if (sizeof (unsigned long) > sizeof (int)) /* param vs _mp_size field */
+     {
+       if (UNLIKELY (new_alloc > INT_MAX))
+ 	{
+ 	  fprintf (stderr, "gmp: overflow in mpz type\n");
+-	  abort ();
++	  __GMP_ALLOC_OVERFLOW_FUNC ();
+ 	}
+     }
+ 
+   PTR(x) = __GMP_ALLOCATE_FUNC_LIMBS (new_alloc);
+   ALLOC(x) = new_alloc;
+diff --git a/mpz/realloc.c b/mpz/realloc.c
+--- a/mpz/realloc.c
++++ b/mpz/realloc.c
+@@ -44,19 +44,19 @@ void *
+   if (sizeof (mp_size_t) == sizeof (int))
+     {
+       if (UNLIKELY (new_alloc > ULONG_MAX / GMP_NUMB_BITS))
+ 	{
+ 	  fprintf (stderr, "gmp: overflow in mpz type\n");
```

Comment:
These print statements aren't ideal. Maybe move them to the default handler function?

Reply:
Good idea; moving to default handler

Reply:
I don't know how to provide a custom error message in jl_throw, any ideas?

Reply:
I don't think we need to; replacing it with a julia exception and stack trace seems sufficient.